File: app-android/app/src/main/java/de/mimuc/senseeverything/db/models/LogDataDao.java
package de.mimuc.senseeverything.db.models;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Update;

import java.util.List;

@Dao
public interface LogDataDao {

    @Query("SELECT * FROM logdata")
    List<LogData> getAll();

    @Insert
    void insertAll(LogData... logDatas);

    @Query("SELECT * FROM logdata WHERE synced = FALSE AND timestamp <= :cutoffTimestamp ORDER BY timestamp ASC LIMIT :n")
    List<LogData> getNextNUnsyncedBefore(int n, long cutoffTimestamp);

    @Query("SELECT COUNT(*) FROM logdata WHERE synced = FALSE")
    long getUnsyncedCount();

    @Query("SELECT COUNT(*) FROM logdata WHERE synced = FALSE AND timestamp <= :cutoffTimestamp")
    long getUnsyncedCountBefore(long cutoffTimestamp);

    @Query("SELECT * FROM logdata ORDER BY timestamp DESC LIMIT 1")
    LogData getLastItem();

    @Update
    public void updateLogData(LogData... logData);

    @Delete
    public void deleteLogData(LogData... logData);

    @Query("DELETE FROM logdata")
    public void deleteAll();
}

File: app-android/app/src/main/java/de/mimuc/senseeverything/db/models/LogDataDao.kt
package de.mimuc.senseeverything.db.models

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.Query
import androidx.room.Update
import kotlinx.coroutines.flow.Flow

@Dao
interface LogDataDao {
    @get:Query("SELECT * FROM logdata")
    val all: List<LogData>

    @Insert
    fun insertAll(vararg logDatas: LogData?)

    @Query("SELECT * FROM logdata WHERE synced = FALSE AND timestamp <= :cutoffTimestamp ORDER BY timestamp ASC LIMIT :n")
    fun getNextNUnsyncedBefore(n: Int, cutoffTimestamp: Long): List<LogData>

    @get:Query("SELECT COUNT(*) FROM logdata WHERE synced = FALSE")
    val unsyncedCount: Long

    @Query("SELECT COUNT(*) FROM logdata WHERE synced = FALSE AND timestamp <= :cutoffTimestamp")
    fun getUnsyncedCountBefore(cutoffTimestamp: Long): Long

    @Query("SELECT COUNT(*) FROM logdata WHERE synced = FALSE AND timestamp <= :cutoffTimestamp")
    fun getUnsyncedCountBeforeFlow(cutoffTimestamp: Long): Flow<Long>

    @get:Query("SELECT * FROM logdata ORDER BY timestamp DESC LIMIT 1")
    val lastItem: LogData?

    @Update
    fun updateLogData(vararg logData: LogData?)

    @Delete
    fun deleteLogData(vararg logData: LogData?)

    @Query("DELETE FROM logdata")
    fun deleteAll()
}
